### ISSUE TITLE
test: Expand integration tests for LossFunctions module (5 missing) [P0]

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#619]
## Overview

This issue tracks the expansion of integration tests for the **LossFunctions** module to achieve complete coverage. This is a P0 (Critical Priority) module.

Parent Issue: #615

## Current Status
- **Coverage:** 85% (29/34 loss functions tested)
- **Source Files:** 34 files (33 loss functions + 1 base class)
- **Test Files:** LossFunctionsIntegrationTests.cs, LossFunctionsMathematicalTests.cs

## Module Location
`src/LossFunctions/`

## Already Tested (29 loss functions)
- BinaryCrossEntropyLoss
- CategoricalCrossEntropyLoss
- ContrastiveLoss
- CosineSimilarityLoss
- CrossEntropyLoss
- DiceLoss
- ElasticNetLoss
- ExponentialLoss
- FocalLoss
- HingeLoss
- HuberLoss
- JaccardLoss
- KullbackLeiblerDivergence
- LogCoshLoss
- MarginLoss
- MeanAbsoluteErrorLoss
- MeanBiasErrorLoss
- MeanSquaredErrorLoss
- ModifiedHuberLoss
- OrdinalRegressionLoss
- PoissonLoss
- QuantileLoss
- RootMeanSquaredErrorLoss
- SparseCategoricalCrossEntropyLoss
- SquaredHingeLoss
- TripletLoss
- WassersteinLoss
- WeightedCrossEntropyLoss

### 1. CTCLoss (Connectionist Temporal Classification)
**Use Case:** Sequence-to-sequence learning without alignment (speech recognition, OCR)
**Tests Needed:**
- [ ] Basic CTC loss computation
- [ ] Test with blank token handling
- [ ] Test with variable-length sequences
- [ ] Compare against PyTorch nn.CTCLoss
- [ ] Test edge cases (empty sequences, single character)
- [ ] Verify gradient computation

### 2. NoiseContrastiveEstimationLoss (NCE)
**Use Case:** Efficient training of language models, word embeddings
**Tests Needed:**
- [ ] Basic NCE loss computation
- [ ] Test with different noise distributions
- [ ] Test with varying number of negative samples
- [ ] Compare against reference implementations
- [ ] Verify probability estimation
- [ ] Verify gradient computation

### 3. PerceptualLoss
**Use Case:** Image generation, style transfer, super-resolution
**Tests Needed:**
- [ ] Basic perceptual loss computation
- [ ] Test with different feature extractors
- [ ] Test with various image sizes
- [ ] Verify feature extraction is correct
- [ ] Test combination with other losses
- [ ] Verify gradient computation

### 4. QuantumLoss
**Use Case:** Quantum machine learning applications
**Tests Needed:**
- [ ] Basic quantum loss computation
- [ ] Test with quantum state vectors
- [ ] Test with density matrices
- [ ] Verify quantum fidelity calculations
- [ ] Verify gradient computation

### 5. RotationPredictionLoss
**Use Case:** Self-supervised learning, image rotation prediction
**Tests Needed:**
- [ ] Basic rotation prediction loss
- [ ] Test with 0, 90, 180, 270 degree rotations
- [ ] Verify classification accuracy
- [ ] Compare against cross-entropy for rotation task
- [ ] Verify gradient computation

### Expand Coverage
- [ ] Add numerical gradient verification for ALL loss functions
- [ ] Add edge case tests (NaN inputs, zero inputs)
- [ ] Add large-scale tests (many samples)
- [ ] Add tests for multi-dimensional outputs

### Cross-Framework Comparison
- [ ] Compare all loss functions against PyTorch equivalents
- [ ] Document any numerical differences

### 1. Basic Computation Tests
- [ ] Verify loss returns expected value for known inputs
- [ ] Verify loss is non-negative where applicable
- [ ] Verify loss is zero for perfect predictions

### 2. Gradient Tests
- [ ] Compare analytical gradient against numerical gradient
- [ ] Verify gradient shape matches input shape
- [ ] Test gradient with edge case inputs

### 3. Edge Cases
- [ ] Test with NaN inputs (should handle gracefully)
- [ ] Test with Infinity inputs
- [ ] Test with zero inputs
- [ ] Test with single sample

### 4. Serialization Tests
- [ ] Save and load loss function configuration
- [ ] Verify behavior matches after reload

### CTCLoss
Forward-backward algorithm on CTC graph. See original paper by Graves et al.

### NCE Loss
```
L = -log(sigmoid(s(w, c))) - sum(log(sigmoid(-s(w, n_i))))
```
where s is score function, c is context, n_i are negative samples.

### Perceptual Loss
```
L = sum(||phi_i(x) - phi_i(y)||^2)
```
where phi_i is feature extractor at layer i.

## Priority Order

1. **High (Test First):**
   - CTCLoss (common in sequence tasks)
   - PerceptualLoss (common in image generation)

2. **Medium:**
   - NoiseContrastiveEstimationLoss
   - RotationPredictionLoss

3. **Lower:**
   - QuantumLoss (specialized use case)

## Acceptance Criteria

- [ ] All 5 missing loss functions have comprehensive tests
- [ ] Numerical gradient verification for new loss functions
- [ ] Edge cases tested
- [ ] Serialization works
- [ ] 100% coverage of LossFunctions module
- [ ] All tests pass on both net8.0 and net471

## References
- PyTorch loss functions: https://pytorch.org/docs/stable/nn.html#loss-functions
- CTC Loss paper: "Connectionist Temporal Classification"
- NCE paper: "Noise-contrastive estimation"
